# Guided Project: Building a Spam Filter with Naive Bayes

### Exploring the Dataset

In this guided project, we're going to study the practical side of the algorithm by building a spam filter for SMS messages.

To classify messages as spam or non-spam, we saw that the computer:

* Learns how humans classify messages.
* Uses that human knowledge to estimate probabilities for new messages — probabilities for spam and non-spam.
* Classifies a new message based on these probability values — if the probability for spam is greater, then it classifies the message as spam. Otherwise, it classifies it as non-spam (if the two probability values are equal, then we may need a human to classify the message).

So our first task is to "teach" the computer how to classify messages. To do that, we'll use the multinomial Naive Bayes algorithm along with a dataset of 5,572 SMS messages that are already classified by humans.

The dataset was put together by Tiago A. Almeida and José María Gómez Hidalgo, and it can be downloaded from the [The UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection). You can also download the dataset directly from this [link](https://dq-content.s3.amazonaws.com/433/SMSSpamCollection). The data collection process is described in more details on [this page](http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/#composition), where you can also find some of the authors' papers.

Note that due to the nature of spam messages, the dataset contains content that may be offensive to some users.

### Read the dataset

- We're going start by reading SMSSpamCollection file using `read_csv()` function from `pandas` package
- The data points are tab separated, so we'll need to use the `sep = '\t'` parameter
- The dataset doesn't have a header row, which means we need to use the `header = None` parameter, otherwise the first row will be wrongly used as the header row.
- And let's use the `names =  ['Label','SMS']` parameter to name the columns as `Label` and `SMS`

In [1]:
import pandas as pd     #importing pandas package

df = pd.read_csv("SMSSpamCollection",sep = "\t",header = None, names = ['Label','SMS'])

In [2]:
df.head()

Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [3]:
df.shape

(5572, 2)

In [4]:
round(df.Label.value_counts(normalize = True)*100,2) 

ham     86.59
spam    13.41
Name: Label, dtype: float64

In [5]:
df.Label.value_counts()

ham     4825
spam     747
Name: Label, dtype: int64

The dataset has 5572 rows and 2 columns. About the 87 % of the total are ham messages (no spam) and the remaining 13 % are spam. Now that we've become more familiar with the dataset, we can move on to build the spam filter.

### Training and Test Set

However, before creating it, it's very helpful to first think of a way of testing how well it works. When creating software (a spam filter is software), a good rule of thumb is that designing the test comes before creating the software. If we write the software first, then it's tempting to come up with a biased test just to make sure the software passes it.

Once our spam filter is done, we'll need to test how good it is with classifying new messages. To test the spam filter, we're first going to split our dataset into two categories:

* A training set, which we'll use to "train" the computer how to classify messages.
* A test set, which we'll use to test how good the spam filter is with classifying new messages.

We're going to keep 80% of our dataset for training, and 20% for testing (we want to train the algorithm on as much data as possible, but we also want to have enough test data). The dataset has 5,572 messages, which means that:

* The training set will have 4,458 messages (about 80% of the dataset).
* The test set will have 1,114 messages (about 20% of the dataset).

To better understand the purpose of putting a test set aside, let's begin by observing that all 1,114 messages in our test set are already classified by a human. When the spam filter is ready, we're going to treat these messages as new and have the filter classify them. Once we have the results, we'll be able to compare the algorithm classification with that done by a human, and this way we'll see how good the spam filter really is.

For this project, our goal is to create a spam filter that classifies new messages with an accuracy greater than 80% — so we expect that more than 80% of the new messages will be classified correctly as spam or ham (non-spam).

In [6]:
#Randomize the dataset
df_random = df.sample(frac = 1, random_state = 1)

#Find index to split
index_split = round(len(df_random)*0.8)

# Training/Test split
training_set = df_random[:index_split].reset_index(drop = True)
test_set = df_random[index_split:].reset_index(drop = True)

print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


Already we split the dataset in the training and test dataset. We'll now analyze the percentage of spam and ham messages in the training and test sets. We expect the percentages to be close to what we have in the full dataset, where about 87% of the messages are ham, and the remaining 13% are spam.

In [7]:
training_set.Label.value_counts(normalize=True)

ham     0.86541
spam    0.13459
Name: Label, dtype: float64

In [8]:
test_set.Label.value_counts(normalize=True)

ham     0.868043
spam    0.131957
Name: Label, dtype: float64

The are good results. Let's move on to clean the dataset.

### Data Cleaning

To calculate all these probabilities, we'll first need to perform a bit of data cleaning to bring the data in a format that will allow us to extract easily all the information we need.

We will remove all the punctuation from the SMS column and transform all the SMS messages to lowercase

In [9]:
# Before cleaning
training_set.head()

Label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [10]:
# After cleaning
training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')
training_set['SMS'] = training_set['SMS'].str.lower()
training_set.head()

C:\Users\Frank\AppData\Local\Temp/ipykernel_1344/3773386248.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')


Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

### Creating the Vocabulary

Our end goal with this data cleaning process is to bring our training set to this format:

![process](naive.png)

With the exception of the "Label" column, every other column in the transformed table above represents a unique word in our vocabulary (more specifically, each column shows the frequency of that unique word for any given message). Recall from the previous lesson that we call the set of unique words a vocabulary.

We'll eventually bring the training set to that format ourselves, but first, let's create a list with all of the unique words that occur in the messages of our training set.

In [11]:
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []

for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))

len(vocabulary)

7783

There are 7783 unique words in the training set

### The Final Training Set

We managed to create the vocabulary for our messages in the training set. Now we're going to use the vocabulary to make the data transformation we need.

In [12]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

In [13]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

forever  4brekkie  waste  hf8  do  female  scenery  8883  tomorw  doin  \
0        0         0      0    0   0       0        0     0       0     0   
1        0         0      0    0   0       0        0     0       0     0   
2        0         0      0    0   0       0        0     0       0     0   
3        0         0      0    0   0       0        0     0       0     0   
4        0         0      0    0   0       0        0     0       0     0   

   ...  degree  yeah  hotel  radiator  concerned  man  checked  liverpool  \
0  ...       0     0      0         0          0    0        0          0   
1  ...       0     0      0         0          0    0        0          0   
2  ...       0     0      0         0          0    0        0          0   
3  ...       0     0      0         0          0    0        0          0   
4  ...       0     0      0         0          0    0        0          0   

   wamma  12hrs  
0      0      0  
1      0      0  
2      0      0  
3      0      0  
4      0      0  

[5 rows x 7783 columns]

Let's concatenate the DataFrame we just built above with the DataFrame containing the training set (this way we'll also have the Label and the SMS columns)

In [14]:
training = pd.concat([training_set, word_counts], axis=1)
training.head()

Label                                                SMS  forever  4brekkie  \
0   ham                  [yep, by, the, pretty, sculpture]        0         0   
1   ham  [yes, princess, are, you, going, to, make, me,...        0         0   
2   ham                    [welp, apparently, he, retired]        0         0   
3   ham                                           [havent]        0         0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...        0         0   

   waste  hf8  do  female  scenery  8883  ...  degree  yeah  hotel  radiator  \
0      0    0   0       0        0     0  ...       0     0      0         0   
1      0    0   0       0        0     0  ...       0     0      0         0   
2      0    0   0       0        0     0  ...       0     0      0         0   
3      0    0   0       0        0     0  ...       0     0      0         0   
4      0    0   0       0        0     0  ...       0     0      0         0   

   concerned  man  checked  liverpool  wamma  12hrs  
0          0    0        0          0      0      0  
1          0    0        0          0      0      0  
2          0    0        0          0      0      0  
3          0    0        0          0      0      0  
4          0    0        0          0      0      0  

[5 rows x 7785 columns]

### Calculating Constants First

Now that we're done with data cleaning and have a training set to work with, we can begin creating the spam filter. Recall that the Naive Bayes algorithm will need to know the probability values of the two equations below to be able to classify new messages:

\begin{equation}
P(Spam | w_1,w_2, ..., w_n) \propto P(Spam) \cdot \prod_{i=1}^{n}P(w_i|Spam) \\
P(Ham | w_1,w_2, ..., w_n) \propto P(Ham) \cdot \prod_{i=1}^{n}P(w_i|Ham)
\end{equation}

Also, to calculate P(wi|Spam) and P(wi|Ham) inside the formulas above, recall that we need to use these equations:

\begin{equation}
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}} \\
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
\end{equation}

Some of the terms in the four equations above will have the same value for every new message. As a start, let's first calculate:
* P(Spam) and P(Ham)
* N<sub>Spam</sub>, N<sub>Ham</sub>, N<sub>Vocabulary</sub>
* Recall from the previous lesson that:

1. N<sub>Spam</sub> is equal to the number of words in all the spam messages — it's not equal to the number of spam messages, and it's not equal to the total number of unique words in spam messages.
2. N<sub>Ham</sub> is equal to the number of words in all the non-spam messages — it's not equal to the number of non-spam messages, and it's not equal to the total number of unique words in non-spam messages.

We'll also use Laplace smoothing and set α = 1

In [15]:
# Isolating spam and ham messages first
spam_messages = training[training['Label'] == 'spam']
ham_messages = training[training['Label'] == 'ham']

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training)
p_ham = len(ham_messages) / len(training)

# N_Spam
n_words_per_spam_message = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam_message.sum()

# N_Ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

# N_Vocabulary
n_vocabulary = len(vocabulary)

# Laplace smoothing
alpha = 1

### Calculating Parameters

The constants parameters are already calculated and we can move on with calculating the parameters **P(w<sub>i</sub>|Ham)** and **P(w<sub>i</sub>|Spam)**. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

We have 7,783 words in our vocabulary, which means we'll need to calculate a total of 15,566 probabilities. For each word, we need to calculate both P(wi|Spam) and P(wi|Ham).

In more technical language, the probability values that P(wi|Spam) and P(wi|Ham) will take are called **parameters**.

\begin{equation}
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}} \\
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
\end{equation}

In [16]:
# Initiate parameters
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()   # spam_messages already defined in a cell above
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()   # ham_messages already defined in a cell above
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha*n_vocabulary)
    parameters_ham[word] = p_word_given_ham

### Classifying A New Message

Now that we've calculated all the constants and parameters we need, we can start creating the spam filter. The spam filter can be understood as a function that:
* Takes in as input a new message (w1, w2, ..., wn)
* Calculates P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn)
* Compares the values of P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn), and:
    - If P(Ham|w1, w2, ..., wn) > P(Spam|w1, w2, ..., wn), then the message is classified as ham.
    - If P(Ham|w1, w2, ..., wn) < P(Spam|w1, w2, ..., wn), then the message is classified as spam.
    - If P(Ham|w1, w2, ..., wn) = P(Spam|w1, w2, ..., wn), then the algorithm may request human help.

In [17]:
import re

def classify(message):
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
            
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal probabilities, have a human classify this!')

In [18]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam


In [19]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham


The function works great with this two messages!!

### Measuring the Spam Filter's Accuracy

We managed to create a spam filter, and we classified two new messages. We'll now try to determine how well the spam filter does on our test set of 1,114 messages.

The algorithm will output a classification label for every message in our test set, which we'll be able to compare with the actual label (given by a human). Note that, in training, our algorithm didn't see these 1,114 messages, so every message in the test set is practically new from the perspective of the algorithm.

First off, we'll change the classify() function that we wrote previously to return the labels instead of printing them.

In [20]:
def classify_test_set(message):

    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()

    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]

        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]

    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

Now that we have a function that returns labels instead of printing them, we can use it to create a new column in our test set.

In [21]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

Now we can compare the predicted values with the actual values to measure how good our spam filter is with classifying new messages. To make the measurement, we'll use **accuracy** as a metric:

\begin{equation}
\text{Accuracy} = \frac{\text{number of correctly classified messages}}{\text{total number of classified messages}}
\end{equation}

In [22]:
%%timeit
correct = 0
total = test_set.shape[0]
    
for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1
#        
#print('Correct:', correct)
#print('Incorrect:', total - correct)
#print('Accuracy:', correct/total)

104 ms ± 2.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


Another way faster :)

In [23]:
total = test_set.shape[0]
incorrect = test_set[test_set["predicted"] != test_set["Label"]].shape[0]
correct = test_set[test_set["predicted"] == test_set["Label"]].shape[0]
accuracy = correct/total
print(f'Correct: {correct:>11}')
print(f'Incorrect: {total - correct:>9}')
print(f'Accuracy: {correct/total*100:>10.2f} %')

Correct:        1100
Incorrect:        14
Accuracy:      98.74 %


### Next Steps

In this project, we managed to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. The filter had an accuracy of 98.74% on the test set, which is an excellent result.